### PR TITLE
Async server implementation

### PR DESCRIPTION
**Pull request by tmcgilchrist**
The implementation is available with this commit 2b13b10a9ff048953cb8689a25cdd22257499a48

It provides a minimal working version of server.ml for Async using Pipes with a greeter example server for testing. However if needs further testing of the other RPC types by implementing the route guide server and testing on real world examples. 

~~This is on top of https://github.com/dialohq/ocaml-grpc/pull/25.~~

### CODE REVIEW

**Comment by quernd** [comment]
Wow, you’re on fire! Thank you for this contribution.
I don’t have enough experience with Async unfortunately. Would you be willing to take a look @mbacarella ?

**Comment by tmcgilchrist** [comment]
Thanks. I’m not sure the types of the Pipe on the streaming type signatures
are quite right. Please hold off on merging while I do some more research.

**Comment by tmcgilchrist** [comment]
I am happier with this version @mbacarella. The types in `server.mli` are more obviously correct now: 
```ocaml
  type unary = string -> (Grpc.Status.t * string option) Deferred.t
  (** [unary] is the type for a unary grpc rpc, one request, one response. *)

  type client_streaming =
    string Pipe.Reader.t -> (Grpc.Status.t * string option) Deferred.t
  (** [client_streaming] is the type for an rpc where the client streams the requests and the server responds once. *)

  type server_streaming =
    string -> string Pipe.Writer.t -> Grpc.Status.t Deferred.t
  (** [server_streaming] is the type for an rpc where the client sends one request and the server sends multiple responses. *)

  type bidirectional_streaming =
    string Pipe.Reader.t -> string Pipe.Writer.t -> Grpc.Status.t Deferred.t
  (** [bidirectional_streaming] is the type for an rpc where both the client and server can send multiple messages. *)
```

There is a route guide implementation using Async that works using `Pipe.t` it might have some mistakes, I haven't used `Pipe.t` before.

File: examples/routeguide-async/proto/dune [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+(rule
+ (targets route_guide.ml)
+ (deps
+  (:proto route_guide.proto))
+ (action
+  (run
+   protoc
+   -I
+   .
+   "--ocaml_out=open=Core;annot=[@@deriving show { with_path = false }, eq, sexp, compare]:."
```
Comment (tmcgilchrist) [inline examples/routeguide-async/proto/dune:17]: Provides an example of deriving sexp and ppx_jane for ocaml-protoc-plugin

**Comment by tmcgilchrist** [comment]
The failure on [debian-12-5.1](https://ocaml.ci.dev/github/dialohq/ocaml-grpc/commit/74110556549fcf0e83b698563078c1d86ac4c433/variant/debian-12-5.1_opam-2.1/-/refs/pull/27/head) should be fixed soon. Core.v0.16.0 is not compatible with OCaml 5.1.

**Pull request merged**